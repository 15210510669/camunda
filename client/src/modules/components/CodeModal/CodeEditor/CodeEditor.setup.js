/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

const initialValueTypes = {
  string: 'SomeString',
  validJSON: '{"firstname":"Max","lastname":"Muster","age":31}',
  array:
    '[{"code":"123.135.625","name":"Laptop Lenovo ABC-001","quantity":1,"price":488.0},{"code":"111.653.365","name":"Headset Sony QWE-23","quantity":2,"price":72.0}]',
  brokenJSON: '"firstname":"Bro","lastname":"Ken","age":31}',
};

const props = {
  contentEditable: true,
  initialValue: initialValueTypes.string,
  handleChange: jest.fn(),
};

const viewerMounts = {
  ...props,
  contentEditable: false,
};

const editorMounts = {
  ...props,
};

const editorMountsWithString = {
  ...props,
  initialValue: initialValueTypes.string,
};

const editorMountsWithNullValue = {
  ...props,
  initialValue: '',
};

const editorMountsWithJSON = {
  ...props,
  initialValue: initialValueTypes.validJSON,
};

const editorMountsWithArray = {
  ...props,
  initialValue: initialValueTypes.string,
};

const editorMountsWithBrokenJSON = {
  ...props,
  initialValue: initialValueTypes.brokenJSON,
};

export const testData = {
  viewerMounts,
  editorMounts,
  editorMountsWithString,
  editorMountsWithJSON,
  editorMountsWithArray,
  editorMountsWithBrokenJSON,
  editorMountsWithNullValue,
};
