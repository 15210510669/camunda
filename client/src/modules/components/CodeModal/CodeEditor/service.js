/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import sanitizeHtml from 'sanitize-html';

export function destructurePasteEvent(event) {
  const content = event.clipboardData.getData('Text');
  const isNewLine = event.target.localName === 'br';
  const targetElement = isNewLine ? event.target.parentElement : event.target;
  return {content, targetElement};
}

export function getHtmlContent(HTMLstring) {
  return sanitizeHtml(HTMLstring, {
    allowedTags: []
  });
}

export function getLocalCaretPosition() {
  const sel = window.getSelection();
  const doesRangeExist = sel.rangeCount !== 0;
  return doesRangeExist ? sel.getRangeAt(0).startOffset : 0;
}

export function getCaretPosition(prevHTML, elementIndex) {
  const sel = window.getSelection();
  const doesRangeExist = sel.rangeCount !== 0;
  const localCaretPosition = doesRangeExist ? sel.getRangeAt(0).startOffset : 0;
  let globalCharsBeforeCaret = 0;
  let currentElementIndex = 0;

  sanitizeHtml(prevHTML, {
    exclusiveFilter: function({text}) {
      if (currentElementIndex === elementIndex) {
        return;
      }
      globalCharsBeforeCaret += text.length;
      currentElementIndex++;
    }
  });

  return globalCharsBeforeCaret + localCaretPosition;
}

export function setBasicCaret(el, targetElementIndex, newCaretPosition) {
  const range = document.createRange();
  const sel = window.getSelection();
  sel.removeAllRanges();

  const element = el.current.children[targetElementIndex];

  if (!element) {
    return;
  }

  // element's first child is the textcontent, if no next -> default caret position
  let {targetElement, position} = element.firstChild
    ? {targetElement: element.firstChild, position: newCaretPosition}
    : {targetElement: element, position: 0};

  // set caret at correct or maximal possible position
  if (targetElement.length < position) {
    position = targetElement.length;
  }

  range.setStart(targetElement, position);
  range.setEnd(targetElement, position);
  sel.addRange(range);
}
