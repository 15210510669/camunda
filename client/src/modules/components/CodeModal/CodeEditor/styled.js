/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import styled, {css} from 'styled-components';
import {Colors, themed, themeStyle} from 'modules/theme';

const LinesSeparator = css`
  &:before {
    content: '';
    position: fixed;
    top: 55px;
    bottom: 0;
    left: 0;
    width: 32px;
    border-right: 1px solid
      ${themeStyle({
        dark: Colors.uiDark02,
        light: Colors.uiLight05
      })};

    background-color: ${themeStyle({
      dark: Colors.uiDark01,
      light: Colors.uiLight04
    })};
  }
`;

export const CodeEditor = themed(styled.div`
  padding: 0;
  position: relative;
  counter-reset: line;
  max-height: calc(100% - 5px);

  ${LinesSeparator}
`);

export const Pre = styled.pre`
  width: fit-content;
  margin: 0;
  min-width: 100%;

  > code > p {
    margin: 3px;
    line-height: 14px;
    color: ${themeStyle({
      dark: 'rgba(255, 255, 255, 0.9)',
      light: Colors.uiLight06
    })};
    font-family: IBMPlexMono;
    font-size: 14px;

    &:before {
      left: 5px;
      position: sticky;
      overflow-x: hidden;
      font-size: 12px;
      box-sizing: border-box;
      text-align: right;
      vertical-align: top;
      line-height: 17px;
      counter-increment: line;
      content: counter(line);
      color: ${themeStyle({
        dark: '#ffffff',
        light: Colors.uiLight06
      })};
      display: inline-block;
      width: 35px;
      opacity: ${themeStyle({
        dark: 0.5,
        light: 0.65
      })};
      padding-right: 13px;
      -webkit-user-select: none;
    }
  }
`;
