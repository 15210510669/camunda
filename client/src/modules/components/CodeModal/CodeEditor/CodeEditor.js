/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useEffect, useState, useRef} from 'react';
import ContentEditable from 'react-contenteditable';

import PropTypes from 'prop-types';

import * as Styled from './styled';

import {isValidJSON} from 'modules/utils';
import {
  createBeautyfiedJSON,
  removeWhiteSpaces,
  removeLineBreaks
} from '../service';

const ref = React.createRef();
const defaultContent =
  '<p className="code-line" data-test="codeline-0" key=0 ></p>';

function CodeEditor({contentEditable, initialValue, handleChange}) {
  const [codeHTML, setCodeHTML] = useState(defaultContent);

  const prevContent = usePrevious(ref);

  function usePrevious(codeRef) {
    if (!codeRef) {
      return '';
    }

    const ref = useRef();
    useEffect(() => {
      ref.current = codeRef.current.textContent;
    });
    return ref.current;
  }

  // Set initial Code modal content
  useEffect(
    () => {
      setCodeHTML(
        renderCodeLines(
          isValidJSON(initialValue)
            ? createBeautyfiedJSON(initialValue, 2)
            : initialValue
        )
      );
    },
    [initialValue]
  );

  // update parent with sanitized content when content changed,
  // layout changes are ignored.
  useEffect(
    () => {
      const {textContent} = ref.current;
      if (prevContent !== textContent) {
        handleChange(
          isValidJSON(textContent)
            ? removeWhiteSpaces(removeLineBreaks(textContent))
            : ''
        );
      }
    },
    [codeHTML]
  );

  function renderCodeLines(initialValue) {
    let htmlString = '';
    initialValue.split('\n').forEach((line, idx) => {
      htmlString =
        htmlString +
        `<p className="code-line" data-test="codeline-${idx}" key=${idx}>${line}</p>`;
    });
    return htmlString;
  }

  function handleOnChange(newValue) {
    setCodeHTML(!newValue ? defaultContent : newValue);
  }

  return (
    <Styled.CodeEditor>
      <Styled.Pre>
        <ContentEditable
          innerRef={ref}
          html={codeHTML}
          disabled={!contentEditable}
          onChange={e => handleOnChange(e.target.value)}
          tagName="code"
        />
      </Styled.Pre>
    </Styled.CodeEditor>
  );
}
export default CodeEditor;

CodeEditor.propTypes = {
  contentEditable: PropTypes.bool,
  initialValue: PropTypes.string,
  handleChange: PropTypes.func
};
