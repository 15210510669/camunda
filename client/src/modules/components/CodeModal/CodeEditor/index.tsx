/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useEffect, useState} from 'react';
import ContentEditable from 'react-contenteditable';

import * as Styled from './styled';

import {isValidJSON} from 'modules/utils';

import {createBeautyfiedJSON} from 'modules/utils/variable';

const codeElementRef = React.createRef();

type Props = {
  contentEditable: boolean;
  initialValue?: string;
};

function CodeEditor({contentEditable, initialValue}: Props) {
  const [codeHTML, setCodeHTML] = useState(returnCodeLine());

  // Set initial Code modal content
  useEffect(() => {
    setCodeHTML(
      renderCodeLines(
        isValidJSON(initialValue)
          ? createBeautyfiedJSON(initialValue, 2)
          : initialValue
      )
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  function returnCodeLine(content = '') {
    return `<p className="code-line">${content}</p>`;
  }

  function renderCodeLines(content: any) {
    let htmlString = '';
    content.split('\n').forEach((lineContent: any) => {
      htmlString = htmlString + returnCodeLine(lineContent);
    });
    return htmlString;
  }

  return (
    <Styled.CodeEditor>
      <Styled.Pre>
        <ContentEditable
          tagName="code"
          data-testid="editable-content"
          html={codeHTML}
          // @ts-expect-error ts-migrate(2769) FIXME: Type 'RefObject<unknown>' is not assignable to typ... Remove this comment to see the full error message
          innerRef={codeElementRef}
          disabled={!contentEditable}
        />
      </Styled.Pre>
    </Styled.CodeEditor>
  );
}
export default CodeEditor;
