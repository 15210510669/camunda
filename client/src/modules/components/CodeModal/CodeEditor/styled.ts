/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import styled, {css} from 'styled-components';

const CodeEditor = styled.div`
  ${({theme}) => {
    const colors = theme.colors.modules.codeModal.codeEditor;

    return css`
      padding: 0;
      position: relative;
      counter-reset: line;
      max-height: calc(100% - 5px);

      &:before {
        content: '';
        position: fixed;
        top: 55px;
        bottom: 0;
        left: 0;
        width: 32px;
        border-right: 1px solid ${colors.borderColor};
        background-color: ${colors.backgroundColor};
      }
    `;
  }}
`;

const Pre = styled.pre`
  ${({theme}) => {
    const colors = theme.colors.modules.codeModal.codeEditor.pre;
    const opacity = theme.opacity.modules.codeModal.codeEditor;

    return css`
      width: fit-content;
      margin: 0;
      min-width: 100%;

      > code > p {
        margin: 3px;
        line-height: 14px;
        color: ${colors.color};
        font-family: IBM Plex Mono;
        font-size: 14px;

        &:before {
          left: 5px;
          position: sticky;
          overflow-x: hidden;
          font-size: 12px;
          box-sizing: border-box;
          text-align: right;
          vertical-align: top;
          line-height: 17px;
          counter-increment: line;
          content: counter(line);
          color: ${theme.colors.text02};
          display: inline-block;
          width: 35px;
          opacity: ${opacity};
          padding-right: 13px;
          -webkit-user-select: none;
        }
      }
    `;
  }}
`;

export {CodeEditor, Pre};
