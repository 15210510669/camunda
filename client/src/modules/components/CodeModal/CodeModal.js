/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useState} from 'react';

import PropTypes from 'prop-types';

import {isValidJSON} from 'modules/utils';
import {createBeautyfiedJSON, removeWhiteSpaces} from './service';

import CodeEditor from './CodeEditor';
import Modal from 'modules/components/Modal';

import * as Styled from './styled';

const MODE = {EDIT: 'edit', VIEW: 'view'};

function CodeModal({
  handleModalClose,
  handleModalSave,
  isModalVisible,
  mode,
  headline,
  initialValue
}) {
  const [newValue, setNewValue] = useState('');

  function onModalClose() {
    setNewValue('');
    handleModalClose();
  }

  function isValueModified() {
    return (
      newValue !==
      removeWhiteSpaces(
        isValidJSON(initialValue)
          ? createBeautyfiedJSON(initialValue)
          : initialValue
      )
    );
  }

  return (
    <Modal onModalClose={onModalClose} isVisible={isModalVisible}>
      <Modal.Header>{headline}</Modal.Header>
      <Styled.ModalBody>
        <CodeEditor
          initialValue={initialValue}
          handleChange={textContent => setNewValue(textContent)}
          contentEditable={mode === MODE.EDIT}
        />
      </Styled.ModalBody>
      <Modal.Footer>
        {mode === MODE.EDIT ? (
          <>
            <Modal.SecondaryButton title="Close Modal" onClick={onModalClose}>
              Close
            </Modal.SecondaryButton>
            <Modal.PrimaryButton
              title="Save Variable"
              data-test="save-btn"
              disableKeyBinding
              disabled={!newValue || !isValueModified()}
              onClick={() => handleModalSave(newValue)}
            >
              Save
            </Modal.PrimaryButton>
          </>
        ) : (
          <Modal.PrimaryButton
            data-test="primary-close-btn"
            title="Close Modal"
            onClick={onModalClose}
          >
            Close
          </Modal.PrimaryButton>
        )}
      </Modal.Footer>
    </Modal>
  );
}

CodeModal.propTypes = {
  handleModalClose: PropTypes.func,
  handleModalSave: PropTypes.func,
  isModalVisible: PropTypes.bool,
  headline: PropTypes.string,
  initialValue: PropTypes.string,
  mode: PropTypes.oneOf(Object.values(MODE)).isRequired
};

export default CodeModal;
