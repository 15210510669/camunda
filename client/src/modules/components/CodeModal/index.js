/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';

import PropTypes from 'prop-types';

import CodeEditor from './CodeEditor';
import Modal, {SIZES} from 'modules/components/Modal';

import * as Styled from './styled';

const MODE = {EDIT: 'edit', VIEW: 'view'};

function CodeModal({
  handleModalClose,
  isModalVisible,
  mode,
  headline,
  initialValue,
}) {
  function onModalClose() {
    handleModalClose();
  }

  return (
    <Modal
      onModalClose={onModalClose}
      isVisible={isModalVisible}
      size={SIZES.BIG}
    >
      <Modal.Header>{headline}</Modal.Header>
      <Styled.ModalBody>
        <CodeEditor
          initialValue={initialValue}
          contentEditable={mode !== MODE.VIEW}
        />
      </Styled.ModalBody>
      <Modal.Footer>
        {mode === MODE.VIEW && (
          <Modal.PrimaryButton
            data-test="primary-close-btn"
            title="Close Modal"
            onClick={onModalClose}
          >
            Close
          </Modal.PrimaryButton>
        )}
      </Modal.Footer>
    </Modal>
  );
}

CodeModal.propTypes = {
  handleModalClose: PropTypes.func,
  isModalVisible: PropTypes.bool,
  headline: PropTypes.string,
  initialValue: PropTypes.string,
  mode: PropTypes.oneOf(Object.values(MODE)).isRequired,
};

export default CodeModal;
