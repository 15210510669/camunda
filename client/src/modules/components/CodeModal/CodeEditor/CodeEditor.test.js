/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {act} from 'react-dom/test-utils';
import {mount} from 'enzyme';

import {testData} from './CodeEditor.setup';

import {ThemeProvider} from 'modules/theme';

import CodeModal from './CodeEditor';

function mountNode(props = {}) {
  return mount(
    <ThemeProvider>
      <CodeModal {...props} />
    </ThemeProvider>
  );
}

describe('CodeModal', () => {
  beforeEach(() => {
    testData.editorMountsWithString.handleChange.mockClear();
  });

  it('should prettify content when initially rendered', () => {
    //given
    const contentTypes = [
      testData.editorMountsWithString,
      testData.editorMountsWithJSON,
      testData.editorMountsWithArray,
      testData.editorMountsWithBrokenJSON,
      testData.editorMountsWithNullValue
    ];

    let node;
    contentTypes.forEach(contentType => {
      //when
      act(() => {
        node = mountNode(contentType);
      });

      // then
      expect(node.find('code').html()).toMatchSnapshot();
    });
  });
});
