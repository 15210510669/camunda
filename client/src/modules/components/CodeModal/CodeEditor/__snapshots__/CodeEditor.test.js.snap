// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeModal should prettify content when initially rendered 1`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\">SomeString</p></code>"`;

exports[`CodeModal should prettify content when initially rendered 2`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\">{</p><p classname=\\"code-line\\" data-test=\\"codeline-1\\" key=\\"1\\">  \\"firstname\\": \\"Max\\",</p><p classname=\\"code-line\\" data-test=\\"codeline-2\\" key=\\"2\\">  \\"lastname\\": \\"Muster\\",</p><p classname=\\"code-line\\" data-test=\\"codeline-3\\" key=\\"3\\">  \\"age\\": 31</p><p classname=\\"code-line\\" data-test=\\"codeline-4\\" key=\\"4\\">}</p></code>"`;

exports[`CodeModal should prettify content when initially rendered 3`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\">SomeString</p></code>"`;

exports[`CodeModal should prettify content when initially rendered 4`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\">\\"firstname\\":\\"Bro\\",\\"lastname\\":\\"Ken\\",\\"age\\":31}</p></code>"`;

exports[`CodeModal should prettify content when initially rendered 5`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\"></p></code>"`;

exports[`CodeModal should render an empty line when no content is passed 1`] = `"<code contenteditable=\\"true\\"><p classname=\\"code-line\\" data-test=\\"codeline-0\\" key=\\"0\\"></p></code>"`;
